### PR TITLE
allow integer as from value in rename

### PR DESCRIPTION
**Pull request by bkamins**
allow integer as `from` value in `rename` and `rename!`. Also clean up the docstrings.

### CODE REVIEW

**Comment by bkamins** [comment]
Fixes #1782

**Comment by kescobo** [comment]
Thank you! I will put this feature to good use 😆

**Comment by bkamins** [comment]
@nalimilan - it would be also good to have this PR for 0.20 as it also cleans up the documentation of things that changed since 0.19.

File: src/abstractdataframe/abstractdataframe.jl [hunk 1]
```diff
@@ -67,6 +67,73 @@ _names(df::AbstractDataFrame) = _names(index(df))
 
 Compat.hasproperty(df::AbstractDataFrame, s::Symbol) = haskey(index(df), s)
 
+"""
+    rename!(df::AbstractDataFrame, vals::AbstractVector{Symbol}; makeunique::Bool=false)
+    rename!(df::AbstractDataFrame, vals::AbstractVector{AbstractString}; makeunique::Bool=false)
+    rename!(df::AbstractDataFrame, (from => to)::Pair...)
+    rename!(df::AbstractDataFrame, d::AbstractDict)
+    rename!(df::AbstractDataFrame, d::AbstractArray{Pair})
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]:
```suggestion
    rename!(df::AbstractDataFrame, d::AbstractArray{<:Pair})
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 2]
```diff
@@ -67,6 +67,73 @@ _names(df::AbstractDataFrame) = _names(index(df))
 
 Compat.hasproperty(df::AbstractDataFrame, s::Symbol) = haskey(index(df), s)
 
+"""
+    rename!(df::AbstractDataFrame, vals::AbstractVector{Symbol}; makeunique::Bool=false)
+    rename!(df::AbstractDataFrame, vals::AbstractVector{AbstractString}; makeunique::Bool=false)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]:
```suggestion
    rename!(df::AbstractDataFrame, vals::AbstractVector{<:AbstractString}; makeunique::Bool=false)
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 3]
```diff
@@ -67,6 +67,73 @@ _names(df::AbstractDataFrame) = _names(index(df))
...
+    rename!(df::AbstractDataFrame, d::AbstractDict)
+    rename!(df::AbstractDataFrame, d::AbstractArray{Pair})
+    rename!(f::Function, df::AbstractDataFrame)
+
+Rename columns of `df` in-place.
+Each name is changed at most once. Permutation of names is allowed.
+
+# Arguments
+- `df` : the `AbstractDataFrame`
+- `d` : an `AbstractDict` type or an `AbstractVector` of `Pairs` that maps
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]:
```suggestion
- `d` : an `AbstractDict` or an `AbstractVector` of `Pair`s that maps
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 4]
```diff
@@ -67,6 +67,73 @@ _names(df::AbstractDataFrame) = _names(index(df))
...
+  the original names or column numbers to new names
+- `f` : a function which for each column takes the old name (a `Symbol`)
+  and returns the new name that gets converted to a `Symbol`
+- `vals` : new column names as a vector of `Symbol`s or `AbstractStrings`
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]:
```suggestion
- `vals` : new column names as a vector of `Symbol`s or `AbstractString`s
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 5]
```diff
@@ -67,6 +67,73 @@ _names(df::AbstractDataFrame) = _names(index(df))
...
+  of the same length as the number of columns in `df`
+- `makeunique` : if `false` (the default), an error will be raised
+  if duplicate names are found; if `true`, duplicate names will be suffixed
+  with `_i` (`i` starting at 1 for the first duplicate).
+
+If pairs are passed to `rename!` (as positional arguments or by a dictionary or a vector)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]:
```suggestion
If pairs are passed to `rename!` (as positional arguments or in a dictionary or a vector)
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 6]
```diff
@@ -67,6 +67,73 @@ _names(df::AbstractDataFrame) = _names(index(df))
...
+then:
+* `from` value can be a `Symbol`, an `AbstractString` or an `Integer`;
+* `to` value can be a `Symbol`, an `AbstractString`.
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]:
```suggestion
* `to` value can be a `Symbol` or an `AbstractString`.
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 7]
```diff
@@ -67,6 +67,73 @@ _names(df::AbstractDataFrame) = _names(index(df))
...
+
+It is required that the type of `from` and of `to` values is the same (e.g.
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: This appears to mean that `from` and `to` must have the same type. Maybe just say "Mixing symbols and strings in `to` and `from` is not allowed."?

File: src/abstractdataframe/abstractdataframe.jl [hunk 8]
```diff
@@ -96,67 +163,107 @@ function rename!(df::AbstractDataFrame,
...
-       makeunique::Bool=false) = rename!(copy(df), vals, makeunique=makeunique)
-rename(df::AbstractDataFrame, args...) = rename!(copy(df), args...)
-rename(f::Function, df::AbstractDataFrame) = rename!(f, copy(df))
-
 """
     rename!(df::AbstractDataFrame, vals::AbstractVector{Symbol}; makeunique::Bool=false)
-    rename!(df::AbstractDataFrame, (from => to)::Pair{Symbol, Symbol}...)
-    rename!(df::AbstractDataFrame, d::AbstractDict{Symbol,Symbol})
-    rename!(df::AbstractDataFrame, d::AbstractArray{Pair{Symbol,Symbol}})
+    rename!(df::AbstractDataFrame, vals::AbstractVector{AbstractString}; makeunique::Bool=false)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: Should be `rename` here and below.

File: src/abstractdataframe/abstractdataframe.jl [hunk 9]
```diff
@@ -67,6 +67,73 @@ _names(df::AbstractDataFrame) = _names(index(df))
...
+julia> rename!(df, [:a, :b, :a])
+ERROR: ArgumentError: Duplicate variable names: :a. Pass makeunique=true to make them unique using a suffix automatically.
+
+julia> rename(df, [:a, :b, :a], makeunique=true)
+1×3 DataFrame
+│ Row │ a     │ b     │ a_1   │
+│     │ Int64 │ Int64 │ Int64 │
+├─────┼───────┼───────┼───────┤
+│ 1   │ 1     │ 2     │ 3     │
+```
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: Add an example with a function while you're at it?

File: src/abstractdataframe/abstractdataframe.jl (hunk 7 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl]: ok - fixed

File: src/abstractdataframe/abstractdataframe.jl (hunk 9 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl]: added (I had it in `rename` but it is ok to add it here also)

File: src/abstractdataframe/abstractdataframe.jl (hunk 8 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl]: fixed

**Comment by bkamins** [comment]
Thank you  for the comments. I have also synced `rename` with `rename!` docs (I must have been tired when writing the old docs - sorry for this).

**Comment by bkamins** [comment]
CI passes, the 'red' is due to coverage

File: src/abstractdataframe/abstractdataframe.jl [hunk 10]
```diff
@@ -132,6 +131,15 @@ julia> rename(df, [:a, :b, :a], makeunique=true)
 │     │ Int64 │ Int64 │ Int64 │
 ├─────┼───────┼───────┼───────┤
 │ 1   │ 1     │ 2     │ 3     │
+
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: Above it should be `rename!`, right?

Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl]: right - fixed

**Pull request merged**